# 3-Manifolds and VOA characters
## Example section companion notebook

The simplest class of examples to consider in verifying the relations between the $\hat{Z}$ invariant and characters of VOA operators is that of Brieskorn spheres. In this example notebook, we replicate the computations done in writing the first paragraph of section 6.1 of arXiv:2201.04640, while at the same time demonstrating the capabilities and main functions of pySeifert: a Sage toolkit for the computation of $\hat{Z}$ invariants for Seifert manifolds and related q-series invariants.

We start by loading pySeifert and defining the Seifert manifold

In [10]:
load("../../pySeifert.sage")
S = Seifert([-1,3,5,2,7,1,9]);S

Seifert manifold with 3 exceptional fibers.
Seifert data:
[-1, 3, 5, 2, 7, 1, 9]
Plumbing Matrix:
[-1  1  0  1  0  1]
[ 1 -2  1  0  0  0]
[ 0  1 -3  0  0  0]
[ 1  0  0 -4  1  0]
[ 0  0  0  1 -2  0]
[ 1  0  0  0  0 -9]
D: 1, m: 315, det(M): 1

The plumbing matrix of $S$ is unitary, therefore $S$ is spherical and we have $D=1$.
The next step in computing the $\hat{Z}$ invariant is to compute the representatives for the boundary conditions, $\mathcal B$, and the $\vec s$ values for the gauge group we have chosen. In this case we focus on $A_2$.

In [11]:
B = S.boundary_conditions(["A",2], basis = "weight"); B # Specifying the basis will help avoid mistakes along the way

[[(-1, -1), (0, 0), (1, 1), (0, 0), (1, 1), (1, 1)]]

As expected, $\mathcal B$ only contains the trivial $\vec{\underline{b}}_0$. With this, we can compute the $\vec{s}$ values.

In [4]:
b0 = B[0]
lVals,sVals,kVals = list(zip(*S.s_values(["A",2],b0, basis = "weight")))
len(sVals),list(zip(lVals,sVals,kVals))[:3]

(216,
 [(-1, (143, 143), (0, 0)), (1, (73, 178), (0, 0)), (1, (178, 73), (0, 0))])

Because the plumbing matrix is unimodular, there are $6^3=216$ admissible $\vec s$ values and all $\vec \kappa_{\hat{w};\vec{\underline{b}}}$ are $(0,0)$ because $D=1$. We can restrict the set of $\vec s$ values to the the range $0<s\leq m$, reducing the number of $\vec s $ values by a factor of the order of the Weyl group (which is 6 for $A_2$).

In [5]:
in_range_sVals = list()
for s in sVals:
    if 0<s[0]<=S.m and 0<s[1]<=S.m:
        in_range_sVals.append(s)
len(in_range_sVals),in_range_sVals[:3]

(36, [(143, 143), (73, 178), (178, 73)])

For each $\vec s$ we can compute the associated $\hat Z$ invariant integrand  $\tilde{\chi}_{\hat{w};\vec{\underline{b}}}$. For example, choosing $\hat w = (1,1,1)$ leads to:

In [6]:
wh = [identity_matrix(2) if d == 1 else matrix(2) for d in S.deg]
chi_tilde = S.chi_tilde(["A",2],wh, b0, 1000, basis="weight"); chi_tilde

(z0^2*z1^2 - z0^4/z1^2 - z1^4/z0^2 + z0^2/z1^4 + z1^2/z0^4 - 1/(z0^2*z1^2))*q^754/(z0*z1 - z0^2/z1 - z1^2/z0 + z0/z1^2 + z1/z0^2 - 1/(z0*z1)) - q^667 + 2*q^524 - 2*q^238 + q^95

Note that in code we adopt the convention that $\hat{w}$ should be of the same length as the dimensions of the plumbing matrix. A proper $\hat{w}$ should have non-zero entries wherever the degree of the node is 1, indicating an end node. 

For each $\vec s$ we can also compute the triplet character. Choosing the same $\hat{w}$ we get:

In [7]:
rho = weyl_vector(["A",2])
mu = 1/sqrt(S.m)*(rho-sVals[0])
trip_ch = triplet_character(["A",2],vector([0,0]),mu,S.m*S.d,3,1000,basis="weight"); trip_ch

(z0^2*z1^2 - z0^4/z1^2 - z1^4/z0^2 + z0^2/z1^4 + z1^2/z0^4 - 1/(z0^2*z1^2))*q^(237169/315)/(z0*z1 - z0^2/z1 - z1^2/z0 + z0/z1^2 + z1/z0^2 - 1/(z0*z1)) - q^(209764/315) + 2*q^(164719/315) - 2*q^(74629/315) + q^(29584/315)

Therem 4.37 shows that these are the same up to a factor of $q^\delta$. This can be verified easily:

In [8]:
expand(chi_tilde - q^S.delta(["A",2])*trip_ch)

0

Finally, we can compute the full $\hat{Z}$ invariant:

In [9]:
S.z_hat(["A",2],b0,300,basis="weight")

6*(2*q^296 - 2*q^294 + 2*q^287 + 2*q^284 - q^283 - 4*q^280 + 2*q^278 + 4*q^272 - 2*q^267 + 2*q^266 - 2*q^262 + 2*q^260 - 2*q^259 + 2*q^254 + 2*q^252 + 4*q^248 - 2*q^238 + 2*q^236 - 2*q^235 - 2*q^232 + 2*q^228 - 2*q^226 - q^219 - 4*q^211 - 2*q^210 - 2*q^202 - 2*q^196 - 2*q^195 + 2*q^191 - 2*q^190 + 2*q^188 - q^187 + 2*q^179 + 4*q^164 + 2*q^159 + 2*q^158 + 2*q^144 - 2*q^142 - 2*q^126 - 2*q^118 + q^95)*q^3